### PR TITLE
Fixes issue where duplicate names map to the same propagation field 

### LINKED ISSUE #852: Handle duplicate references to the same extra field
In sleuth 2.x `TraceAutoConfiguration`, there's code like this:

```java
		if (!sleuthProperties.getBaggageKeys().isEmpty()) {
			factoryBuilder = factoryBuilder
					// for HTTP
					.addPrefixedFields("baggage-", sleuthProperties.getBaggageKeys())
					// for messaging
					.addPrefixedFields("baggage_", sleuthProperties.getBaggageKeys());
		}
		if (!sleuthProperties.getPropagationKeys().isEmpty()) {
			for (String key : sleuthProperties.getPropagationKeys()) {
				factoryBuilder = factoryBuilder.addField(key);
			}
		}
```

The following input will result in 3 values for the same keys:
```
spring.sleuth.baggage-keys=userId,sessionId
spring.sleuth.propagation-keys=userId,sessionId
```

Specifically, it will:
* add "baggage-" prefixed "userId" and "sessionId" via `factoryBuilder.addPrefixedFields`
* add "baggage_" prefixed "userId" and "sessionId" via `factoryBuilder.addPrefixedFields`
* add "userId" and "sessionId" via `factoryBuilder.addField`

When I first looked at this, I noticed the redundant `baggage_` which will still break on messaging for certain keys due to problems mentioned in #826. However, I think even two redundant additions of the same keys (ex "userId" and "sessionId") would be a problem in setting and getting the values.

Eventhough it is probably quite bad overhead to add the same properties 3 times, we should probably not make it worse by having the implementation not find a single correct value per key.

Thanks @tdanylchuk for bringing this up

### PR DESCRIPTION
**Pull request by adriancole**
Fixes #852

### CODE REVIEW

File: brave/src/main/java/brave/propagation/ExtraFieldPropagation.java [hunk 1]
```diff
@@ -72,7 +74,7 @@
   public static Factory newFactory(Propagation.Factory delegate, String... fieldNames) {
     if (delegate == null) throw new NullPointerException("delegate == null");
     if (fieldNames == null) throw new NullPointerException("fieldNames == null");
-    String[] validated = ensureLowerCase(Arrays.asList(fieldNames));
+    String[] validated = ensureLowerCase(new LinkedHashSet<>(Arrays.asList(fieldNames)));
```
Comment (adriancole) [inline brave/src/main/java/brave/propagation/ExtraFieldPropagation.java]: needs a test

File: brave/src/main/java/brave/propagation/ExtraFieldPropagation.java [hunk 2]
```diff
@@ -81,7 +83,7 @@ public static Factory newFactory(Propagation.Factory delegate,
       Collection<String> fieldNames) {
     if (delegate == null) throw new NullPointerException("delegate == null");
     if (fieldNames == null) throw new NullPointerException("fieldNames == null");
-    String[] validated = ensureLowerCase(fieldNames);
+    String[] validated = ensureLowerCase(new LinkedHashSet<>(fieldNames));
```
Comment (adriancole) [inline brave/src/main/java/brave/propagation/ExtraFieldPropagation.java]: needs a test

File: brave/src/main/java/brave/propagation/ExtraFieldPropagation.java [hunk 3]
```diff
@@ -123,7 +125,7 @@ public FactoryBuilder addPrefixedFields(String prefix, Collection<String> fieldN
       if (prefix == null) throw new NullPointerException("prefix == null");
       if (prefix.isEmpty()) throw new IllegalArgumentException("prefix is empty");
       if (fieldNames == null) throw new NullPointerException("fieldNames == null");
-      prefixedNames.put(prefix, ensureLowerCase(fieldNames));
+      prefixedNames.put(prefix, ensureLowerCase(new LinkedHashSet<>(fieldNames)));
```
Comment (adriancole) [inline brave/src/main/java/brave/propagation/ExtraFieldPropagation.java]: needs a test

File: brave/src/main/java/brave/propagation/ExtraFieldPropagation.java [hunk 4]
```diff
@@ -133,20 +135,38 @@ public Factory build() {
         return new Factory(delegate, validated, validated);
       }
       List<String> fields = new ArrayList<>(), keys = new ArrayList<>();
+      ;
```
Comment (adriancole) [inline brave/src/main/java/brave/propagation/ExtraFieldPropagation.java]: whoops

File: brave/src/main/java/brave/propagation/ExtraFieldPropagation.java [hunk 5]
```diff
@@ -222,10 +242,23 @@ public static void set(TraceContext context, String name, String value) {
...
     final String[] fieldNames;
     final String[] keyNames;
+    final int[] keyToField;
     final ExtraFactory extraFactory;
 
     Factory(Propagation.Factory delegate, String[] fieldNames, String[] keyNames) {
+      this(delegate, fieldNames, keyNames, keyToField(keyNames));
+    }
+
+    static int[] keyToField(String[] keyNames) {
```
Comment (adriancole) [inline brave/src/main/java/brave/propagation/ExtraFieldPropagation.java]: somewhere there should be a javadoc saying:

"we have a key to field mapping as there may be multiple propagation keys that reference the same field. For example, baggage-userid and baggage_userid"

**Comment by adriancole** [comment]
ok should be good to go now

**Comment by tdanylchuk** [comment]
can we merge fix?

**Comment by adriancole** [comment]
yep. thanks for the reminder

**Comment by tdanylchuk** [comment]
awesome! looking forward for release in order to be committed into sleuth and test it=)

**Comment by adriancole** [comment]
mind testing snapshot meanwhile with the brave-bom? snapshots are
hosted on jfrog

**Pull request merged**